### PR TITLE
Support for 64-bit integers

### LINKED ISSUE #446: Need for long integers fo rdiagnostics?
Primarily in diagnostics, like for origin of parcels. Probably not needed for main code.

### PR DESCRIPTION
**Pull request by matt-frey**
Add support for 64-bit integers.

Closes #446.

### CODE REVIEW

File: mpi-tests/benchmark_parcel_merging.f90 [hunk 1]
```diff
@@ -0,0 +1,228 @@
...
+            do
+                call get_command_argument(i, arg)
+                if (len_trim(arg) == 0) then
+                    exit
+                endif
+
+                if (arg == '--nx') then
+                    i = i + 1
+                    call get_command_argument(i, arg)
+                    read(arg,'(i6)') nx
```
Comment (sjboeing) [inline mpi-tests/benchmark_parcel_merging.f90:148]: what does i6 mean here?

File: src/mpi/mpi_collectives.f90 [hunk 2]
```diff
@@ -37,4 +50,152 @@ subroutine mpi_integer_reduce(sendbuf, op, comm)
...
+
+                call mpi_check_for_error(cart, &
+                    "in MPI_Reduce_local of mpi_neighbor_allreduce_logical.")
+            enddo
+
+        end subroutine mpi_neighbor_allreduce_logical
+
+        !::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::
+
+        subroutine mpi_neighbor_allreduce_integer(sendbuf, recvbuf, op)
```
Comment (sjboeing) [inline src/mpi/mpi_collectives.f90:131]: Very minor, I think we have mixed spelling (neighbor/neighbour)

**Review by sjboeing — COMMENTED** [review]
I have gone through this now. It is hard for me to understand everything, but I think we want to merge this sometime soon so that not too much else happens. Is there anything in particular where you would like me to check details?

File: src/3d/parcels/parcel_nearest.f90 [hunk 3]
```diff
@@ -824,26 +902,28 @@ subroutine resolve_tree(isma, iclo, rclo, n_local_small)
                             !     INTEGER, OPTIONAL, INTENT(OUT) :: ierror
                             l_helper = .true.
                             offset = ic - 1 ! starts at 0
-                            call MPI_Rput(l_helper,         &
+                            call MPI_Put(l_helper,          &
```
[the remark this answers is not in the document]
Comment (matt-frey) [reply src/3d/parcels/parcel_nearest.f90:905]: I changed to MPI_Put as I think it is better to use here.

**Comment by matt-frey** [comment]
I think it would be good you to have a look at the benchmark examples in detail, i.e. are you happy with the parcel setup etc.

**Comment by sjboeing** [comment]
I had a look at this, it looks good to me.

File: src/mpi/mpi_collectives.f90 (hunk 2 above)
Comment (matt-frey) [reply src/mpi/mpi_collectives.f90]: Removed.

**Review by sjboeing — APPROVED** [review]
Thanks Matthias. I had another quick glance at this and nothing stands out to me. Feel free to merge.

**Pull request merged**